File: frontend/src/services/aiService.ts
import api from "./api";

/* -----------------------------
   Health
------------------------------ */

export interface AIHealthResponse {
  status: string;
  provider: string;
  model: string;
}

/* -----------------------------
   Chat
------------------------------ */

export interface ChatRequest {
  message: string;
  language: "Tamil" | "English";
}

export interface ChatResponse {
  reply: string;
  language: string;
}

/* -----------------------------
   Explain
------------------------------ */

export interface ExplainRequest {
  topic: string;
  language: "Tamil" | "English";
}

export interface ExplainResponse {
  title: string;
  explanation: string;
  key_points: string[];
  important_facts: string[];
  memory_trick?: string;
}

/* -----------------------------
   MCQ Generator
------------------------------ */

export interface MCQRequest {
  topic: string;
  difficulty: "Easy" | "Medium" | "Hard";
  number_of_questions: number;
}

export interface MCQOption {
  option: string;
}

export interface MCQ {
  question: string;
  options: MCQOption[];
  answer: string;
  explanation: string;
}

export interface MCQResponse {
  topic: string;
  questions: MCQ[];
}

/* -----------------------------
   Study Planner
------------------------------ */

export interface StudyPlanRequest {
  exam: string;
  days_left: number;
  study_hours_per_day: number;
}

export interface StudyTask {
  day: number;
  subject: string;
  topic: string;
  duration_hours: number;
}

export interface StudyPlanResponse {
  exam: string;
  total_days: number;
  tasks: StudyTask[];
}

/* -----------------------------
   Performance Analysis
------------------------------ */

export interface PerformanceAnalysisRequest {
  weak_subjects: string[];
  average_score: number;
  target_exam: string;
}

export interface PerformanceAnalysisResponse {
  readiness_percentage: number;
  weak_topics: string[];
  recommendations: string[];
  revision_priority: string[];
}

/* -----------------------------
   Summary
------------------------------ */

export interface SummaryRequest {
  title: string;
  content: string;
  language: "Tamil" | "English";
}

export interface SummaryResponse {
  title: string;
  summary: string;
  important_points: string[];
}

/* -----------------------------
   AI Service
------------------------------ */

export const AIService = {

  async health(): Promise<AIHealthResponse> {
    const { data } = await api.get("/ai/health");
    return data;
  },

  async chat(
    request: ChatRequest,
  ): Promise<ChatResponse> {

    const { data } = await api.post(
      "/ai/chat",
      request,
    );

    return data;
  },

  async explain(
    request: ExplainRequest,
  ): Promise<ExplainResponse> {

    const { data } = await api.post(
      "/ai/explain",
      request,
    );

    return data;
  },

  async generateMCQ(
    request: MCQRequest,
  ): Promise<MCQResponse> {

    const { data } = await api.post(
      "/ai/generate-mcq",
      request,
    );

    return data;
  },

  async studyPlan(
    request: StudyPlanRequest,
  ): Promise<StudyPlanResponse> {

    const { data } = await api.post(
      "/ai/study-plan",
      request,
    );

    return data;
  },

  async analyzePerformance(
    request: PerformanceAnalysisRequest,
  ): Promise<PerformanceAnalysisResponse> {

    const { data } = await api.post(
      "/ai/analyze-performance",
      request,
    );

    return data;
  },

  async summarize(
    request: SummaryRequest,
  ): Promise<SummaryResponse> {

    const { data } = await api.post(
      "/ai/summarize",
      request,
    );

    return data;
  },

};

File: backend/app/services/ai_service.py
from app.schemas.ai import (
    ChatRequest,
    ChatResponse,
    ExplainRequest,
    ExplainResponse,
    MCQRequest,
    MCQResponse,
    MCQ,
    MCQOption,
    StudyPlanRequest,
    StudyPlanResponse,
    StudyTask,
    PerformanceAnalysisRequest,
    PerformanceAnalysisResponse,
    SummaryRequest,
    SummaryResponse,
    AIHealthResponse,
)


class AIService:
    """
    AI Service Layer

    NOTE:
    This version is provider-independent.
    Later you can connect:
    - OpenAI
    - Google Gemini
    - Azure OpenAI
    - Ollama (Offline)
    - HuggingFace
    """

    # -------------------------------------------------
    # Health
    # -------------------------------------------------

    def health(self) -> AIHealthResponse:

        return AIHealthResponse(
            status="online",
            provider="Local",
            model="TNPSC MASTER AI",
        )

    # -------------------------------------------------
    # Chat
    # -------------------------------------------------

    def chat(
        self,
        request: ChatRequest,
    ) -> ChatResponse:

        return ChatResponse(
            reply=(
                f"AI response for: {request.message}\n\n"
                "Connect an AI provider (Gemini/OpenAI/Ollama) "
                "to generate intelligent answers."
            ),
            language=request.language,
        )

    # -------------------------------------------------
    # Explain Topic
    # -------------------------------------------------

    def explain(
        self,
        request: ExplainRequest,
    ) -> ExplainResponse:

        return ExplainResponse(
            title=request.topic,
            explanation=(
                f"This is a placeholder explanation for "
                f"'{request.topic}'."
            ),
            key_points=[
                "Important Point 1",
                "Important Point 2",
                "Important Point 3",
            ],
            important_facts=[
                "TNPSC Important Fact",
            ],
            memory_trick="Remember using simple keywords.",
        )

    # -------------------------------------------------
    # Generate MCQs
    # -------------------------------------------------

    def generate_mcq(
        self,
        request: MCQRequest,
    ) -> MCQResponse:

        questions = []

        for i in range(request.number_of_questions):

            questions.append(

                MCQ(
                    question=f"{request.topic} Question {i+1}",
                    options=[
                        MCQOption(option="Option A"),
                        MCQOption(option="Option B"),
                        MCQOption(option="Option C"),
                        MCQOption(option="Option D"),
                    ],
                    answer="Option A",
                    explanation="Placeholder explanation.",
                )

            )

        return MCQResponse(
            topic=request.topic,
            questions=questions,
        )

    # -------------------------------------------------
    # Study Planner
    # -------------------------------------------------

    def study_plan(
        self,
        request: StudyPlanRequest,
    ) -> StudyPlanResponse:

        tasks = []

        for day in range(
            1,
            request.days_left + 1,
        ):

            tasks.append(

                StudyTask(
                    day=day,
                    subject="General Studies",
                    topic=f"Revision {day}",
                    duration_hours=request.study_hours_per_day,
                )

            )

        return StudyPlanResponse(
            exam=request.exam,
            total_days=request.days_left,
            tasks=tasks,
        )

    # -------------------------------------------------
    # Performance Analysis
    # -------------------------------------------------

    def analyze_performance(
        self,
        request: PerformanceAnalysisRequest,
    ) -> PerformanceAnalysisResponse:

        return PerformanceAnalysisResponse(
            readiness_percentage=request.average_score,
            weak_topics=request.weak_subjects,
            recommendations=[
                "Practice daily.",
                "Revise weak subjects.",
                "Take mock tests.",
            ],
            revision_priority=request.weak_subjects,
        )

    # -------------------------------------------------
    # Summary
    # -------------------------------------------------

    def summarize(
        self,
        request: SummaryRequest,
    ) -> SummaryResponse:

        summary = request.content[:500]

        return SummaryResponse(
            title=request.title,
            summary=summary,
            important_points=[
                "Important Point 1",
                "Important Point 2",
                "Important Point 3",
            ],
        )